feat: [CDS-70736]: Add OCI support, ECR AWS configuration (#2179)
harness/developer-hub: docs/continuous-delivery/gitops/helm-oci-repository-aws-ecr.md

---
title: Setup access to OCI Helm repository in private AWS ECR
description: This topic describes how can we rotate token for accessing AWS ECR where we can store OCI Helm repository.
sidebar_position: 8

> helpdocs_topic_id: lf6a27usso

Author: how is this populated? @roshnisarangadharan

Reviewer: You only need the first 3 items. Ignore the rest. They're part of old topics.

---
title: Store and access OCI Helm repository in private Amazon ECR
description: This topic describes how to store OCI Helm repository in private Amazon ECR, and access them by rotating tokens.
sidebar_position: 8
---

> This topic assumes that you have already created as [OCI Helm repository](/docs/continuous-delivery/gitops/add-a-harness-git-ops-repository#add-a-repository).

Reviewer 2: ```suggestion
This topic assumes that you have already created an [OCI Helm repository](/docs/continuous-delivery/gitops/add-a-harness-git-ops-repository#add-a-repository).
```

> When storing an OCI Helm repository in Amazon Elastic Container Registry (Amazon ECR), you must obtain a token. The Amazon ECR tokens are short-lived, hence need rotation. 

Reviewer 2: ```suggestion
When storing an OCI Helm repository in Amazon Elastic Container Registry (ECR), you must obtain a token. The Amazon ECR tokens are short-lived, hence need rotation. 
```

> GitOps repositories are stored in Kubernetes secrets. For GitOps to be able to pull these repositories from Amazon ECR, you have to rotate the credentials stored in the secret. You can use [External Secrets Operator](https://external-secrets.io) and [ArgoCD ECR Updater](https://artifacthub.io/packages/helm/argocd-aws-ecr-updater/argocd-ecr-updater) to achieve this.

Reviewer: ```suggestion
GitOps repository credentials are stored in Kubernetes secrets. For GitOps to be able to pull these repositories from Amazon ECR, you have to rotate the credentials stored in the secret. You can use [External Secrets Operator](https://external-secrets.io) and [ArgoCD ECR Updater](https://artifacthub.io/packages/helm/argocd-aws-ecr-updater/argocd-ecr-updater) to achieve this.
```

> GitOps repositories are stored in Kubernetes secrets. For GitOps to be able to pull these repositories from Amazon ECR, you have to rotate the credentials stored in the secret. You can use [External Secrets Operator](https://external-secrets.io) and [ArgoCD ECR Updater](https://artifacthub.io/packages/helm/argocd-aws-ecr-updater/argocd-ecr-updater) to achieve this.

Reviewer 2: "GitOps repositories are stored in Kubernetes secrets." 
This doesn't seem correct. Should it be "GitOps repository credentials are stored in Kubernetes secrets"? Please double check with dev.

> In this topic, we will walk you through how to use the external secrets operator.

Reviewer 2: ```suggestion
In this topic, we will walk you through how to use the External Secrets Operator.
```

Reviewer 2: Since you're talking about a third party product, you should capitalize External Secrets Operator throughout the topic.

> 

Reviewer 2: It would be good to have a summary of (numbered) steps here to link to the sections below it.

## Install external secrets in the cluster where Argo CD is installed

```bash
helm repo add external-secrets https://charts.external-secrets.io
helm install external-secrets \
external-secrets/external-secrets \
-n external-secrets \
--create-namespace \
--set installCRDs=true
```

> The above commands installs the external secret operator with necessary Custom Resource Definitions (CRDs). 

Reviewer 2: ```suggestion
The above commands install the external secret operator with necessary Custom Resource Definitions (CRDs). 
```

Reviewer 2: It's best practice to explain what the commands do before presenting them, so this sentence should be above the code snippet.

   
The external secret operator has `ClusterSecretStore` and `ExternalSecret` CRDs that are used to manage the secrets update. `ClusterSecretStore` defines secret storage, whereas `ExternalSecret` connects the Kubernetes secret's storage and destination.
   

> There is an additional CRD, `ECRAuthorizationToken` that generates the token. We will use this token instead of `ClusterSecretStore` in this setup.

Reviewer: ```suggestion
There is an additional CRD, `ECRAuthorizationToken` that generates the token. We will use this CRD instead of `ClusterSecretStore` in this setup.
```

## Create AWS credentials in Kubernetes secret
   

> Use the following command to create AWS credentials in Kubernetes secret: 

Reviewer 2: ```suggestion
Use the following commands to create AWS credentials in Kubernetes secret: 
```

```bash
export AWS_ACCESS_KEY_ID=AKIAIOSFODNN7EXAMPLE
export AWS_SECRET_ACCESS_KEY=wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY

echo -n AWS_SECRET_ACCESS_KEY > ./secret-access-key
echo -n AWS_ACCESS_KEY_ID  > ./access-key  
kubectl create secret generic awssm-secret --from-file=./access-key  --from-file=./secret-access-key
```

> The credentials used below are samples only. Use your actual credentials.

Reviewer 2: ```suggestion
The credentials used above are samples only. Use your actual credentials.
```

Reviewer 2: ideally, this sentence should be above the code snippet, and in that case, it would point to "below".

> The credentials used below are samples only. Use your actual credentials.

Reviewer: ```suggestion
The credentials used above are samples only. Use your actual credentials.
```

> ## Create Generator YAML

Reviewer 2: ```suggestion
## Create generator YAML
```

> The credentials created in the previous step `awssm-secret` is referenced here.

Reviewer 2: ```suggestion
The secret created in the previous step, `awssm-secret`, is referenced below.
```

   
```yaml
apiVersion: generators.external-secrets.io/v1alpha1
kind: ECRAuthorizationToken
metadata:
  name: ecr-gen
spec:

  # specify aws region (mandatory)
  region: us-west-1

  auth:

    # 1: static credentials
    # point to a secret that contains static credentials
    # like AWS_ACCESS_KEY_ID / AWS_SECRET_ACCESS_KEY
    secretRef:
      accessKeyIDSecretRef:
        name: "awssm-secret"
        key: "access-key"
      secretAccessKeySecretRef:
        name: "awssm-secret"
        key: "secret-access-key"
```

> ## Apply Generator YAML

Reviewer 2: ```suggestion
## Apply generator YAML
```
